File: VB/DatabaseUserSettings/DatabaseUserSettingsStore.vb
Imports System.IO
Imports DevExpress.ExpressApp
Imports DevExpress.Data.Filtering
Imports DevExpress.Persistent.Base
Imports DevExpress.ExpressApp.Utils
Imports DevExpress.ExpressApp.Model
Imports DevExpress.ExpressApp.Model.Core

Namespace DatabaseUserSettings
    Public Enum ImportUserSettingsBehavior
        Merge
        Overwrite
    End Enum
    Public Interface IDatabaseUserSettings
        ReadOnly Property Aspects() As IList(Of IDatabaseUserSettingsAspect)
        Property UserId() As String
        ReadOnly Property UserName() As String
        Sub Reset()
        Sub Export()
        Sub ImportFrom(ByVal source As IDatabaseUserSettings, ByVal importBehavior As ImportUserSettingsBehavior)
    End Interface
    Public Interface IDatabaseUserSettingsAspect
        Property Name() As String
        Property Owner() As IDatabaseUserSettings
        Property Xml() As String
    End Interface
    Public Interface IManageDatabaseUserSettingsParameter
        Property Source() As IDatabaseUserSettings
        ReadOnly Property Targets() As IList(Of IDatabaseUserSettings)
        Property ImportBehavior() As ImportUserSettingsBehavior
    End Interface
    Public Class DatabaseUserSettingsStore
        Inherits ModelDifferenceStore

        Private _application As XafApplication
        Private [module] As DatabaseUserSettingsModule
        Public Const XafmlHeader As String = "<?xml version=""1.0"" encoding=""utf-8""?>"
        Public Const XafmlRootElement As String = "<Application/>"
        Public Shared ReadOnly EmptyXafml As String = String.Format("{0}{1}{2}", DatabaseUserSettingsStore.XafmlHeader, Environment.NewLine, DatabaseUserSettingsStore.XafmlRootElement)
        Public Sub New(ByVal application As XafApplication)
            Me._application = application
            Guard.ArgumentNotNull(application, "application")
            [module] = CType(Me.Application.Modules.FirstOrDefault(Function(m) m.GetType() Is GetType(DatabaseUserSettingsModule)), DatabaseUserSettingsModule)
            Guard.ArgumentNotNull([module], "module")
        End Sub
        Public Overrides ReadOnly Property Name() As String
            Get
                Return Me.GetType().Name
            End Get
        End Property
        Protected ReadOnly Property Application() As XafApplication
            Get
                Return _application
            End Get
        End Property
        Protected ReadOnly Property DatabaseUserSettingsModule() As DatabaseUserSettingsModule
            Get
                Return [module]
            End Get
        End Property
        Public Shared Function GetUserSettings(ByVal objectSpace As IObjectSpace, ByVal userSettingsType As Type, ByVal userId As String) As IDatabaseUserSettings
            Dim userSettings As IDatabaseUserSettings = Nothing
            If String.IsNullOrEmpty(userId) Then
                userSettings = GetConfiguratorUserSettings(objectSpace, userSettingsType)
            Else
                Dim user As Object = objectSpace.FindObject(DatabaseUserSettingsModule.UserTypeInfo.Type, New OperandProperty(DatabaseUserSettingsModule.UserTypeInfo.KeyMember.Name) = New OperandValue(DatabaseUserSettingsModule.UserIdTypeConverter.ConvertFromInvariantString(userId)))
                If user IsNot Nothing Then
                    If Convert.ToString(DatabaseUserSettingsModule.UserTypeInfo.FindMember(DatabaseUserSettingsModule.UserNameProperty).GetValue(user)) = DatabaseUserSettingsModule.ConfiguratorUserName Then
                        userSettings = GetConfiguratorUserSettings(objectSpace, userSettingsType)
                    Else
                        userSettings = GetUserUserSettings(objectSpace, userSettingsType, userId)
                    End If
                End If
            End If
            Return userSettings
        End Function
        Private Shared Function GetUserUserSettings(ByVal objectSpace As IObjectSpace, ByVal userSettingsType As Type, ByVal userId As String) As IDatabaseUserSettings
            Dim userSettings As IDatabaseUserSettings = CType(objectSpace.FindObject(userSettingsType, New OperandProperty(DatabaseUserSettingsModule.UserIdProperty) = New OperandValue(userId)), IDatabaseUserSettings)
            If userSettings Is Nothing Then
                userSettings = CreateUserSettings(objectSpace, userSettingsType)
                userSettings.UserId = userId
            End If
            Return userSettings
        End Function
        Public Shared Function GetConfiguratorUserSettings(ByVal objectSpace As IObjectSpace, ByVal userSettingsType As Type) As IDatabaseUserSettings
            Dim userSettings As IDatabaseUserSettings = CType(objectSpace.FindObject(userSettingsType, New NullOperator(DatabaseUserSettingsModule.UserIdProperty)), IDatabaseUserSettings)
            If userSettings Is Nothing Then
                userSettings = CreateUserSettings(objectSpace, userSettingsType)
                Dim fileUserSettingsStore As New FileModelStore(PathHelper.GetApplicationFolder(), AppDiffDefaultName)
                Dim aspects As IEnumerable(Of String) = fileUserSettingsStore.GetAspects().Concat(New String() { String.Empty })
                For Each aspectName As String In aspects
                    Try
                        GetUserSettingsAspect(objectSpace, userSettings, aspectName).Xml = File.ReadAllText(PathHelper.GetApplicationFolder() & fileUserSettingsStore.GetFileNameForAspect(aspectName))
                    Catch e As Exception
                        Tracing.Tracer.LogError(e)
                        Throw New UserFriendlyException(Localization.CannotLoadUserSettingsFromFile & Localization.UserSettingsFailureSuggestion & Environment.NewLine & e.Message)
                    End Try
                Next aspectName
            End If
            Return userSettings
        End Function
        Public Overrides Sub Load(ByVal model As ModelApplicationBase)
            Try
                Using objectSpace As IObjectSpace = Application.CreateObjectSpace()
                    Dim reader As New ModelXmlReader()
                    Dim userSettings As IDatabaseUserSettings = GetUserSettings(objectSpace, DatabaseUserSettingsModule.UserSettingsType, DatabaseUserSettingsModule.UserIdTypeConverter.ConvertToInvariantString(SecuritySystem.CurrentUserId))
                    For Each aspect As IDatabaseUserSettingsAspect In userSettings.Aspects
                        If String.IsNullOrEmpty(aspect.Xml) Then
                            aspect.Xml = EmptyXafml
                        End If
                        reader.ReadFromString(model, aspect.Name, aspect.Xml)
                    Next aspect
                    objectSpace.CommitChanges()
                End Using
            Catch e As Exception
                If Not(TypeOf e Is InvalidOperationException) Then
                    Tracing.Tracer.LogError(e)
                    Throw New UserFriendlyException(Localization.CannotLoadUserSettingsFromTheDatabase & Localization.UserSettingsFailureSuggestion & Environment.NewLine & e.Message)
                End If
            End Try
        End Sub
        Public Overrides Sub SaveDifference(ByVal model As ModelApplicationBase)
            Try
                Using objectSpace As IObjectSpace = Application.CreateObjectSpace()
                    Dim userSettings As IDatabaseUserSettings = GetUserSettings(objectSpace, DatabaseUserSettingsModule.UserSettingsType, DatabaseUserSettingsModule.UserIdTypeConverter.ConvertToInvariantString(SecuritySystem.CurrentUserId))
                    For i As Integer = 0 To model.AspectCount - 1
                        Dim writer As New ModelXmlWriter()
                        Dim xml As String = writer.WriteToString(model, i)
                        If Not String.IsNullOrEmpty(xml) Then
                            Dim aspectName As String = model.GetAspect(i)
                            Dim aspect As IDatabaseUserSettingsAspect = GetUserSettingsAspect(objectSpace, userSettings, aspectName)
                            aspect.Xml = String.Format("{0}{1}{2}", XafmlHeader, Environment.NewLine, xml)
                        End If
                    Next i
                    objectSpace.CommitChanges()
                End Using
            Catch e As Exception
                If Not(TypeOf e Is InvalidOperationException) Then
                    Tracing.Tracer.LogError(e)
                    Throw New UserFriendlyException(Localization.CannotSaveUserSettingsToTheDatabase & Localization.UserSettingsFailureSuggestion & Environment.NewLine & e.Message)
                End If
            End Try
        End Sub
        Public Shared Function FindAspectByName(ByVal userSettings As IDatabaseUserSettings, ByVal aspectName As String) As IDatabaseUserSettingsAspect
            If userSettings Is Nothing OrElse userSettings.Aspects Is Nothing Then
                Return Nothing
            End If
            Dim userSettingsAspect As IDatabaseUserSettingsAspect = userSettings.Aspects.FirstOrDefault(Function(a) a.Name = aspectName)
            Return userSettingsAspect
        End Function
        Public Shared Function GetUserSettingsAspect(ByVal objectSpace As IObjectSpace, ByVal userSettings As IDatabaseUserSettings, ByVal aspectName As String) As IDatabaseUserSettingsAspect
            If userSettings Is Nothing Then
                Return Nothing
            End If
            Dim userSettingsAspect As IDatabaseUserSettingsAspect = FindAspectByName(userSettings, aspectName)
            If userSettingsAspect Is Nothing Then
                userSettingsAspect = CreateUserSettingsAspect(objectSpace, objectSpace.TypesInfo.FindTypeInfo(CType(userSettings, Object).GetType()).FindMember("Aspects").ListElementType)
                userSettingsAspect.Owner = userSettings
                userSettingsAspect.Name = aspectName
            End If
            Return userSettingsAspect
        End Function
        Private Shared Function CreateUserSettings(ByVal objectSpace As IObjectSpace, ByVal userSettingsType As Type) As IDatabaseUserSettings
            Return CType(objectSpace.CreateObject(userSettingsType), IDatabaseUserSettings)
        End Function
        Private Shared Function CreateUserSettingsAspect(ByVal objectSpace As IObjectSpace, ByVal modelAspectType As Type) As IDatabaseUserSettingsAspect
            Return CType(objectSpace.CreateObject(modelAspectType), IDatabaseUserSettingsAspect)
        End Function
    End Class
End Namespace

File: CS/DatabaseUserSettings/DatabaseUserSettingsStore.cs
using System;
using System.IO;
using System.Linq;
using DevExpress.ExpressApp;
using DevExpress.Data.Filtering;
using DevExpress.Persistent.Base;
using System.Collections.Generic;
using DevExpress.ExpressApp.Utils;
using DevExpress.ExpressApp.Model;
using DevExpress.ExpressApp.Model.Core;

namespace DatabaseUserSettings {
    public enum ImportUserSettingsBehavior { Merge, Overwrite };
    public interface IDatabaseUserSettings {
        IList<IDatabaseUserSettingsAspect> Aspects { get; }
        string UserId { get; set; }
        string UserName { get; }
        void Reset();
        void Export();
        void ImportFrom(IDatabaseUserSettings source, ImportUserSettingsBehavior importBehavior);
    }
    public interface IDatabaseUserSettingsAspect {
        string Name { get; set; }
        IDatabaseUserSettings Owner { get; set; }
        string Xml { get; set; }
    }
    public interface IManageDatabaseUserSettingsParameter {
        IDatabaseUserSettings Source { get; set; }
        IList<IDatabaseUserSettings> Targets { get; }
        ImportUserSettingsBehavior ImportBehavior { get; set; }
    }
    public class DatabaseUserSettingsStore : ModelDifferenceStore {
        private XafApplication application;
        private DatabaseUserSettingsModule module;
        public const string XafmlHeader = "<?xml version=\"1.0\" encoding=\"utf-8\"?>";
        public const string XafmlRootElement = "<Application/>";
        public static readonly string EmptyXafml = String.Format("{0}{1}{2}", DatabaseUserSettingsStore.XafmlHeader, Environment.NewLine, DatabaseUserSettingsStore.XafmlRootElement);
        public DatabaseUserSettingsStore(XafApplication application) {
            this.application = application;
            Guard.ArgumentNotNull(application, "application");
            module = (DatabaseUserSettingsModule)Application.Modules.FirstOrDefault<ModuleBase>(
                m => m.GetType() == typeof(DatabaseUserSettingsModule)
            );
            Guard.ArgumentNotNull(module, "module");
        }
        public override string Name { get { return GetType().Name; } }
        protected XafApplication Application { get { return application; } }
        protected DatabaseUserSettingsModule DatabaseUserSettingsModule { get { return module; } }
        public static IDatabaseUserSettings GetUserSettings(IObjectSpace objectSpace, Type userSettingsType, string userId) {
            IDatabaseUserSettings userSettings = null;
            if (string.IsNullOrEmpty(userId))
                userSettings = GetConfiguratorUserSettings(objectSpace, userSettingsType);
            else {
                object user = objectSpace.FindObject(DatabaseUserSettingsModule.UserTypeInfo.Type, new OperandProperty(DatabaseUserSettingsModule.UserTypeInfo.KeyMember.Name) == new OperandValue(DatabaseUserSettingsModule.UserIdTypeConverter.ConvertFromInvariantString(userId)));
                if (user != null)
                    if (Convert.ToString(DatabaseUserSettingsModule.UserTypeInfo.FindMember(DatabaseUserSettingsModule.UserNameProperty).GetValue(user)) == DatabaseUserSettingsModule.ConfiguratorUserName)
                        userSettings = GetConfiguratorUserSettings(objectSpace, userSettingsType);
                    else
                        userSettings = GetUserUserSettings(objectSpace, userSettingsType, userId);
            }
            return userSettings;
        }
        private static IDatabaseUserSettings GetUserUserSettings(IObjectSpace objectSpace, Type userSettingsType, string userId) {
            IDatabaseUserSettings userSettings = (IDatabaseUserSettings)objectSpace.FindObject(
                userSettingsType,
                new OperandProperty(DatabaseUserSettingsModule.UserIdProperty) == new OperandValue(userId)
            );
            if (userSettings == null) {
                userSettings = CreateUserSettings(objectSpace, userSettingsType);
                userSettings.UserId = userId;
            }
            return userSettings;
        }
        public static IDatabaseUserSettings GetConfiguratorUserSettings(IObjectSpace objectSpace, Type userSettingsType) {
            IDatabaseUserSettings userSettings = (IDatabaseUserSettings)objectSpace.FindObject(
                userSettingsType,
                new NullOperator(DatabaseUserSettingsModule.UserIdProperty)
            );
            if (userSettings == null) {
                userSettings = CreateUserSettings(objectSpace, userSettingsType);
                FileModelStore fileUserSettingsStore = new FileModelStore(PathHelper.GetApplicationFolder(), AppDiffDefaultName);
                IEnumerable<string> aspects = fileUserSettingsStore.GetAspects().Concat(new string[] { string.Empty });
                foreach (string aspectName in aspects)
                    try {
                        GetUserSettingsAspect(objectSpace, userSettings, aspectName).Xml = File.ReadAllText(PathHelper.GetApplicationFolder() + fileUserSettingsStore.GetFileNameForAspect(aspectName));
                    } catch (Exception e) {
                        Tracing.Tracer.LogError(e);
                        throw new UserFriendlyException(Localization.CannotLoadUserSettingsFromFile + Localization.UserSettingsFailureSuggestion + Environment.NewLine + e.Message);
                    }
            }
            return userSettings;
        }
        public override void Load(ModelApplicationBase model) {
            try {
                using (IObjectSpace objectSpace = Application.CreateObjectSpace()) {
                    ModelXmlReader reader = new ModelXmlReader();
                    IDatabaseUserSettings userSettings = GetUserSettings(objectSpace, DatabaseUserSettingsModule.UserSettingsType, DatabaseUserSettingsModule.UserIdTypeConverter.ConvertToInvariantString(SecuritySystem.CurrentUserId));
                    foreach (IDatabaseUserSettingsAspect aspect in userSettings.Aspects) {
                        if (string.IsNullOrEmpty(aspect.Xml))
                            aspect.Xml = EmptyXafml;
                        reader.ReadFromString(model, aspect.Name, aspect.Xml);
                    }
                    objectSpace.CommitChanges();
                }
            } catch (Exception e) {
                if (!(e is InvalidOperationException)) {
                    Tracing.Tracer.LogError(e);
                    throw new UserFriendlyException(Localization.CannotLoadUserSettingsFromTheDatabase + Localization.UserSettingsFailureSuggestion + Environment.NewLine + e.Message);
                }
            }
        }
        public override void SaveDifference(ModelApplicationBase model) {
            try {
                using (IObjectSpace objectSpace = Application.CreateObjectSpace()) {
                    IDatabaseUserSettings userSettings = GetUserSettings(objectSpace, DatabaseUserSettingsModule.UserSettingsType, DatabaseUserSettingsModule.UserIdTypeConverter.ConvertToInvariantString(SecuritySystem.CurrentUserId));
                    for (int i = 0; i < model.AspectCount; i++) {
                        ModelXmlWriter writer = new ModelXmlWriter();
                        string xml = writer.WriteToString(model, i);
                        if (!string.IsNullOrEmpty(xml)) {
                            string aspectName = model.GetAspect(i);
                            IDatabaseUserSettingsAspect aspect = GetUserSettingsAspect(objectSpace, userSettings, aspectName);
                            aspect.Xml = String.Format("{0}{1}{2}", XafmlHeader, Environment.NewLine, xml);
                        }
                    }
                    objectSpace.CommitChanges();
                }
            } catch (Exception e) {
                if (!(e is InvalidOperationException)) {
                    Tracing.Tracer.LogError(e);
                    throw new UserFriendlyException(Localization.CannotSaveUserSettingsToTheDatabase + Localization.UserSettingsFailureSuggestion + Environment.NewLine + e.Message);
                }
            }
        }
        public static IDatabaseUserSettingsAspect FindAspectByName(IDatabaseUserSettings userSettings, string aspectName) {
            if (userSettings == null || userSettings.Aspects == null)
                return null;
            IDatabaseUserSettingsAspect userSettingsAspect = userSettings.Aspects.FirstOrDefault<IDatabaseUserSettingsAspect>(
                a => a.Name == aspectName
            );
            return userSettingsAspect;
        }
        public static IDatabaseUserSettingsAspect GetUserSettingsAspect(IObjectSpace objectSpace, IDatabaseUserSettings userSettings, string aspectName) {
            if (userSettings == null)
                return null;
            IDatabaseUserSettingsAspect userSettingsAspect = FindAspectByName(userSettings, aspectName);
            if (userSettingsAspect == null) {
                userSettingsAspect = CreateUserSettingsAspect(objectSpace, objectSpace.TypesInfo.FindTypeInfo(userSettings.GetType()).FindMember("Aspects").ListElementType);
                userSettingsAspect.Owner = userSettings;
                userSettingsAspect.Name = aspectName;
            }
            return userSettingsAspect;
        }
        private static IDatabaseUserSettings CreateUserSettings(IObjectSpace objectSpace, Type userSettingsType) {
            return (IDatabaseUserSettings)objectSpace.CreateObject(userSettingsType);
        }
        private static IDatabaseUserSettingsAspect CreateUserSettingsAspect(IObjectSpace objectSpace, Type modelAspectType) {
            return (IDatabaseUserSettingsAspect)objectSpace.CreateObject(modelAspectType);
        }
    }
}